packages/haunted-hooks-testing-library/src/index.ts

```diff
@@ -14,36 +16,171 @@ class Result<T> {
-    update(options.initialProps);
+    // options can only be undefined if P is undefined
+    update(_options?.initialProps);
   }, null);
 
-  function update(props) {
+  function update(props?: P) {
     state.run(() => {
-      result.current = hook(props);
+      const res = hook(props);
+      result.current = res;
+      updateSubject.next(res);
     });
   }
 
-  function rerender(newProps?: any) {
-    update(newProps);
+  function rerender(newProps?: P) {
+    if (newProps === undefined) {
+      update(_options?.initialProps);
+    } else {
+      update(newProps);
+    }
     state.runLayoutEffects();
     state.runEffects();
   }
-  rerender(options.initialProps);
+  // options can only be undefined if P is undefined
+  rerender(_options?.initialProps);
+
+  async function waitForNextUpdate(options?: {
+    timeout?: number | false;
+  }): Promise<void> {
+    const timeoutMs: number | false = options?.timeout ?? 1000;
+
+    const updateObservable = timeoutMs
+      ? updateSubject.pipe(timeout(timeoutMs))
+      : updateSubject;
+
+    return updateObservable.pipe(first(), mapTo(undefined)).toPromise();
+  }
+
+  async function waitFor(
+    callback: () => boolean | void,
+    options?: {
+      interval?: number | false;
+      timeout?: number | false;
+    }
+  ): Promise<void> {
+    // should resolve promise if callback returns truthy or undefined
+    // callback can error
+    function shouldResolve() {
+      try {
+        const res = callback();
+        return res === undefined || !!res;
+      } catch {
+        return false;
+      }
+    }
+
+    const timeoutMs: number | false = options?.timeout ?? 1000;
+    const intervalMs: number | false = options?.interval ?? 50;
+
+    const updateObservable = timeoutMs
+      ? updateSubject.pipe(timeout(timeoutMs))
+      : updateSubject;
+
+    const intervalObservable = intervalMs ? interval(intervalMs) : NEVER;
+
+    const mergedPromise = merge(updateObservable, intervalObservable)
+      .pipe(filter(shouldResolve), first(), mapTo(undefined))
+      .toPromise();
+
+    return mergedPromise;
+  }
+
+  async function waitForValueToChange(
+    selector: () => any,
+    options?: {
+      interval?: number | false;
+      timeout?: number | false;
+    }
+  ): Promise<void> {
+    const initialValue = selector();
+
+    const timeoutMs: number | false = options?.timeout ?? 1000;
+    const intervalMs: number | false = options?.interval ?? 50;
+
+    const updateObservable = timeoutMs
+      ? updateSubject.pipe(timeout(timeoutMs))
+      : updateSubject;
+
+    const intervalObservable = intervalMs ? interval(intervalMs) : NEVER;
+
+    const mergedPromise = merge(updateObservable, intervalObservable)
+      .pipe(
+        filter(() => selector() !== initialValue),
+        first(),
+        mapTo(undefined)
+      )
+      .toPromise();
+
+    return mergedPromise;
+  }
 
   return {
     result,
     rerender,
     unmount: () => state.teardown(),
```

Comment:
This should probably also clean up subject.

When does `unmount` get called? What stops the hooks from running infinitely in background?

```suggestion
    unmount: () => { 
        state.teardown()
        // TODO: Teardown subject
    },
```

Reply:
I'll call `.unsubscribe` and `.complete` on the Subject which should stop everything.

`unmount` must be called manually in order for the hook to unmount. I assumed that the hook would be garbage collected when the results of `renderHook` go out of scope, not sure if this is true though.